### PR TITLE
[quant] Add per block quantization primitives

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
We want to use this to replace all quantize/dequantize/choose_qparams ops in https://github.com/pytorch-labs/ao/blob/main/torchao/quantization/quant_primitives.py and https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/fx/_decomposed.py

Note: this PR only adds the ops, we'll do replacement in separate PRs and make sure it does not degrade the performance or accuracy

Test Plan:
python test/quantization/test_quant_primitives.py

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torchao/quantization/quant_primitives.py [hunk 1]
```diff
@@ -41,6 +43,231 @@
...
+            shape_for_reduction.append(input_size[i])
+            cur_dim += 1
+        else:
+            shape_for_reduction.append(block_size[i])
+            reduction_dims.append(cur_dim)
+            cur_dim += 1
+    return shape_for_reduction, reduction_dims
+
+
+def quantize_affine_per_block(
```
Comment (cpuhrsch) [inline torchao/quantization/quant_primitives.py:103]: I think it might be easier to first write a version of this that assumes

`input.dim() == len(block_size) == scale.dim() == zero_point.dim()` and then use various tools to implement broadcasting. But our ops should kind of imply the broadcasting here.

File: torchao/quantization/quant_primitives.py [hunk 2]
```diff
@@ -41,6 +43,245 @@
...
+        torch.uint1: (0, 2**1-1),
+        torch.uint2: (0, 2**2-1),
+        torch.uint3: (0, 2**3-1),
+        torch.uint4: (0, 2**4-1),
+        torch.uint5: (0, 2**5-1),
+        torch.uint6: (0, 2**6-1),
+        torch.uint7: (0, 2**7-1),
+    })
+
+def _get_qmin_qmax(dtype, quant_min, quant_max):
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:65]: I feel this is more like `_check_qmin_qmax`? Alternatively make `quant_min` and `quant_max` default to None?

File: torchao/quantization/quant_primitives.py [hunk 3]
```diff
@@ -41,6 +43,245 @@
...
+    input = input.view(shape_for_reduction)
+    shape_after_reduction = shape_for_reduction
+    for i in reduction_dims:
+        shape_after_reduction[i] = 1
+    scale = scale.view(shape_after_reduction)
+    if zero_point is not None:
+        zero_point = zero_point.view(shape_after_reduction)
+
+    quant = torch.clamp(
+        torch.round(input / scale) + zero_point, quant_min, quant_max
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:166]: I think this is slightly different from existing quant primitives, e.g. `quantize_per_token` does round after adding zp: https://github.com/pytorch/pytorch/blob/d40774f4ed4a45c70d49e66f4e1f197dfc274758/torch/ao/quantization/fx/_decomposed.py#L771

However, as written this is consistent with the existing `torch.fake_quantize_per_channel_affine`, which adds the zp after round. Which one do we want to follow?

File: test/quantization/test_quant_primitives.py [hunk 4]
```diff
@@ -46,5 +53,157 @@ def test_get_group_qparams_symmetric(self):
...
+        quant_max = 127
+        quantized_ref = torch.ops.quantized_decomposed.quantize_per_channel_group(
+            input, scale, zero_point, quant_min, quant_max, torch.int8, group_size
+        )
+        dequantized_ref = torch.ops.quantized_decomposed.dequantize_per_channel_group(
+            quantized_ref, scale, zero_point, quant_min, quant_max, torch.int8, group_size, output_dtype=torch.float32
+        )
+
+        self.assertTrue(torch.equal(quantized, quantized_ref))
+        self.assertTrue(torch.equal(dequantized, dequantized_ref))
```
Comment (andrewor14) [inline test/quantization/test_quant_primitives.py:142]: Does this tell you how many elements were different and by how much? Should we use this instead?
```
torch.testing.assert_close(quantized, quantized_ref, atol=0, rtol=0)
```

File: torchao/quantization/quant_primitives.py [hunk 5]
```diff
@@ -41,6 +43,245 @@
...
+        "quant_max out of bound for dtype, " \
+        f"quant_max_upper_bound: {quant_max_upper_bound} quant_max: {quant_max}"
+    return quant_min, quant_max
+
+def _get_reduction_params(block_size, input_size):
+    """Given block_size and input size find the parameters for reduction:
+
+    Output:
+        shape_for_reduction: the shape we use to `view` input to prepare it for reduction
+        reduction_dims: the dims we'll do reduction over
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:92]: Can you add an example here of what output values we expect given a specific `block_size` and `input_size`? It will help me better understand what this function is doing

File: torchao/quantization/quant_primitives.py [hunk 6]
```diff
@@ -41,6 +43,245 @@
...
+   zero_point_dtype: Optional[torch.dtype] = None,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    Args:
+        input (torch.Tensor): fp32, bf16, fp16 input Tensor
+        mapping_type (MappingType): determines how the qparams are calculated, symmetric or asymmetric
+        dtype (torch.dtype): dtype for target quantized Tensor
+        quant_min (Optional[int]): minimum quantized value for target quantized Tensor
+        quant_max (Optioanl[int]): maximum quantized value for target quantized Tensor
+        eps (Optional[flaot]: minimum scale
```
Comment (andrewor14) [inline torchao/quantization/quant_primitives.py:244]: float

**Review by andrewor14 — COMMENTED** [review]
Looks great! Is the next step to rewrite existing quant primitives using this? How will this work for qdq ops currently living in pytorch?

**Review by cpuhrsch — APPROVED** [review]
Accepting now with the intent of iterating on this over time

File: torchao/quantization/quant_primitives.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:65]: yeah this actually combined two functions, i can split them as well

File: torchao/quantization/quant_primitives.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:166]: I think we should just choose one, we can start with this and check to see if we can adjust others to use this I think, we could make the dtypes more explicit as well

File: torchao/quantization/quant_primitives.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:65]: we are thinking about just don't allow quant_min/quant_max, will update this after we made a decision here, I'll add a TODO here

**Comment by jerryzh168** [comment]
> Looks great! Is the next step to rewrite existing quant primitives using this? How will this work for qdq ops currently living in pytorch?

yeah we'll rewrite existing primitives in torchao to use this first, and then expand to pytorch later, we'll need to move the ops to pytorch in order to refactor the ops there

File: torchao/quantization/quant_primitives.py (hunk 5 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:93]: done

**Comment by jerryzh168** [comment]
I also tried to not include block_size as args and use `keep_dim` for scales, but there was a problem, e.g. when we have:
input: (3, 3, 10, 10)
block_size: (3, 3, 2, 10)

scale size: (1, 1, 5, 1)

I'm not sure how can we broadcast the scale to be size (1, 1, 10, 1) in order for it to be divided by input

**Review by HDCharles — CHANGES_REQUESTED** [review]
these look significantly more complicated. Do these still torch.compile into performant kernels?

Would like to see some ad-hoc microbenchmarks at least to indicate that we're not immediately going to see a huge perf hit from this change

File: torchao/quantization/quant_primitives.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py]: @cpuhrsch any ideas on how we can support broadcasting for the example I described here: https://github.com/pytorch/ao/pull/159#issuecomment-2073887750

File: torchao/quantization/quant_primitives.py [hunk 7]
```diff
@@ -41,6 +43,255 @@
...
+        Output:
+          shape_for_reduction: (3, 3, 5, 2, 10)
+          reduction_dim: [0, 1, 3, 4]
+    """
+    assert len(block_size) == len(input_size)
+    shape_for_reduction = []
+    reduction_dims = []
+    cur_dim = 0
+    for i in range(len(block_size)):
+        if block_size[i] != input_size[i] and block_size[i] > 1:
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:109]: what is `block_size[i] != input_size[i] and block_size[i] == 1`. As in if corresponding block size is 1. What would that mean?

File: torchao/quantization/quant_primitives.py [hunk 8]
```diff
@@ -41,6 +43,255 @@
...
+            assert input_size[i] % block_size[i] == 0, f"Expecting input size at {i} dimension: {input_size[i]} to be divisible by block_size at {i} dimension: {block_size[i]}"
+            shape_for_reduction.append(input_size[i] // block_size[i])
+            shape_for_reduction.append(block_size[i])
+            # reduce over the block_size[i] dim
+            reduction_dims.append(cur_dim + 1)
+            cur_dim += 2
+        else:
+            # block_size[i] == input_size[i] or block_size[i] == 1
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:117]: ok i see it here

File: torchao/quantization/quant_primitives.py [hunk 9]
```diff
@@ -41,6 +43,255 @@
...
+            # otherwise it's already the same as reduced dimension
+            if block_size[i] != 1:
+                reduction_dims.append(cur_dim)
+            cur_dim += 1
+    return shape_for_reduction, reduction_dims
+
+
+def quantize_affine(
+    input: torch.Tensor,
+    block_size: List[int],
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:129]: I thought we decided that you can use scale/zero point shape to infer htis?

File: torchao/quantization/quant_primitives.py [hunk 10]
```diff
@@ -41,6 +43,255 @@
...
+    scale: torch.Tensor,
+    zero_point: Optional[torch.Tensor],
+    output_dtype: torch.dtype,
+    quant_min: Optional[int] = None,
+    quant_max: Optional[int] = None
+):
+    """
+    Args:
+      input (torch.Tensor): original float32 or bfloat16 Tensor
+      block_size: (List[int]): granularity of quantization, this means the size of the tensor elements that's sharing the same qparam
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:139]: I think this is probably better explained with examples.

File: torchao/quantization/quant_primitives.py [hunk 11]
```diff
@@ -41,6 +43,255 @@
...
+      output_dtype (torch.dtype): requested dtype (e.g. torch.uint8) for output Tensor
+      quant_min (Optional[int]): minimum quantized value for output Tensor, if not specified, it will be derived from dtype
+      quant_max (Optional[int]): maximum quantized value for output Tensor, if not specified, it will be derived from dtype
+
+    Note:
+      How can block_size represent different granularities?
+      let's say we have a Tensor of size: (3, 3, 10, 10), here is the table showing how block_size represents different
+      granularities:
+
+       granularity type       |     block_size
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:152]: which I see here. so all good

File: torchao/quantization/quant_primitives.py [hunk 12]
```diff
@@ -41,6 +43,255 @@
...
+     per_group (groupsize=2)  |    (3, 3, 10, 2)
+     per_group (groupsize=2) for axis = 3 | (3, 3, 2, 10)
+
+
+    Output:
+      quantized tensor with requested dtype
+    """
+    # TODO: validations
+    quant_min, quant_max = _get_and_check_qmin_qmax(output_dtype, quant_min, quant_max)
+    shape_for_reduction, reduction_dims = _get_reduction_params(block_size, input.size())
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:165]: should you also validate that the blocksize should also correspond to the scale/zp size?

File: torchao/quantization/quant_primitives.py [hunk 13]
```diff
@@ -41,6 +43,255 @@
...
+    for i in reduction_dims:
+        shape_after_reduction[i] = 1
+    scale = scale.view(shape_after_reduction)
+    if zero_point is not None:
+        zero_point = zero_point.view(shape_after_reduction)
+
+    dequant = input.to(torch.float32)
+    scale = scale.to(torch.float32)
+    if zero_point is not None:
+        zero_point = zero_point.to(torch.float32)
```
Comment (kimishpatel) [inline torchao/quantization/quant_primitives.py:225]: This does not feel accurate. I think we have had some discussion around this that it should be `(input.to(torch.int32) - zero_point.to(torch.int32)).to(torch.float32) * scale)

**Review by kimishpatel — COMMENTED** [review]
Sorry for reviewing htis late so just leaving some comments. Mainly two
- Do we need block size or it can be drived?
- My concern around dequantize routine

File: torchao/quantization/quant_primitives.py (hunk 7 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:109]: block_size[i] == 1 means that for ith dimension, each slice will have their own qparams

File: torchao/quantization/quant_primitives.py (hunk 9 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:129]: there is some issues with broadcasting: https://github.com/pytorch/ao/pull/159#issuecomment-2073887750 let me know if you have some ideas

File: torchao/quantization/quant_primitives.py [hunk 14]
```diff
@@ -41,6 +43,255 @@
...
+    block_size: List[int],
+    scale: torch.Tensor,
+    zero_point: Optional[torch.Tensor],
+    output_dtype: torch.dtype,
+    quant_min: Optional[int] = None,
+    quant_max: Optional[int] = None
+):
+    """
+    Args:
+      input (torch.Tensor): original float32 or bfloat16 Tensor
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:138]: yeah I can assert it's float32, bfloat16 and float16 I think, I'll add this in a future PR

File: torchao/quantization/quant_primitives.py (hunk 12 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:165]: yeah I can add that

File: torchao/quantization/quant_primitives.py (hunk 13 above)
Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:225]: I see, will fix

**Comment by jerryzh168** [comment]
> Sorry for reviewing htis late so just leaving some comments. Mainly two
> 
> * Do we need block size or it can be drived?
> * My concern around dequantize routine

sorry just saw the comments

* yeah blocksize can be derived from some helper functions I think, we could add these when we start replacing callsites I feel
* will fix

**Comment by kimishpatel** [comment]
> I also tried to not include block_size as args and use `keep_dim` for scales, but there was a problem, e.g. when we have: input: (3, 3, 10, 10) block_size: (3, 3, 2, 10)
> 
> scale size: (1, 1, 5, 1)
> 
> I'm not sure how can we broadcast the scale to be size (1, 1, 10, 1) in order for it to be divided by input

input : (3, 3, 10, 10)
scale: (1, 1, 5, 1)

It we assume scale is always of valid shape than to broadcast scale[2] == 5 to input[2]==10, we will have to interpret scale as blockwise scale where block size = 2., Right?
